lib/parameter.js

```diff
@@ -0,0 +1,774 @@
+  if (typeof value === 'boolean') return ParameterType.PARAMETER_BOOL;
+  if (typeof value == 'string') return ParameterType.PARAMETER_STRING;
+  if (typeof value == 'number') return ParameterType.PARAMETER_DOUBLE;
+  if (Array.isArray(value)) {
+    if (value.length > 0) {
+      const elementType = parameterTypeFromValue(value[0]);
+      switch (elementType) {
+      case ParameterType.PARAMETER_BOOL:   return ParameterType.PARAMETER_BOOL_ARRAY;
+      case ParameterType.PARAMETER_DOUBLE: return ParameterType.PARAMETER_DOUBLE_ARRAY;
+      case ParameterType.PARAMETER_STRING: return ParameterType.PARAMETER_STRING_ARRAY;
+      }
+    }
+
+    return ParameterType.PARAMETER_NOT_SET;
+  }
+
+  // otherwise unrecognized value
+  throw new TypeError('Unrecognized parameter type.');
+}
+
+/**
+ * Determine if a number maps to is a valid ParameterType.
+ * 
+ * @param {ParameterType} parameterType - The value to test.
+ * @return {boolean} - True if value is a valid ParameterType; false otherwise.
+ */
+function validType(parameterType) {
+    
+  let result = typeof parameterType === 'number' && 
+      (ParameterType.PARAMETER_NOT_SET <= parameterType <= ParameterType.PARAMETER_STRING_ARRAY);
+  
+  return result;
+}
+
+/**
+ * Test if value can be represented by a ParameterType.
+ * 
+ * @param {number} value - The value to test.
+ * @param {ParameterType} type - The ParameterType to test value against.
+ * @return {boolean} - True if value can be represented by type.
+ */
+function validValue(value, type) {
+
+  if (value == null) {
+    return type === ParameterType.PARAMETER_NOT_SET;
+  }
+
+  let result = true;
+
+  switch (type) {
+  case ParameterType.PARAMETER_NOT_SET:
+    result = !value;
+    break;
+  case ParameterType.PARAMETER_BOOL:
+    result = typeof value == 'boolean';
+    break;
+  case ParameterType.PARAMETER_STRING:
+    result = typeof value == 'string';
+    break;
+  case ParameterType.PARAMETER_INTEGER:
+  case ParameterType.PARAMETER_DOUBLE:
+    result = typeof value == 'number';
+    break;
+  case ParameterType.PARAMETER_BOOL_ARRAY:
+  case ParameterType.PARAMETER_BYTE_ARRAY:
+  case ParameterType.PARAMETER_INTEGER_ARRAY:
+  case ParameterType.PARAMETER_DOUBLE_ARRAY:
+    const values = value;
+
+    result = Array.isArray(values);
+
+    if (result) {
+      let arrayElementType;
+
+      if (type == ParameterType.PARAMETER_BOOL_ARRAY) {
+        arrayElementType = ParameterType.PARAMETER_BOOL;
+      } else if (type == ParameterType.PARAMETER_BYTE_ARRAY) {
+        arrayElementType = ParameterType.PARAMETER_BYTE;
+      } if (type == ParameterType.PARAMETER_INTEGER_ARRAY) {
+        arrayElementType = ParameterType.PARAMETER_INTEGER;
+      } if (type == ParameterType.PARAMETER_DOUBLE_ARRAY) {
+        arrayElementType = ParameterType.PARAMETER_DOUBLE;
+      }
+
+      result = 
+        values.reduce((compatible, val) => 
+          compatible && validValue(val, arrayElementType), 
+        true);
+    }
```

Comment:
Would you please put it into a private function?

Reply:
done, refactored to _isArrayParameterType(type)